spec.bs

```diff
@@ -2877,6 +2891,38 @@ CORP violation report=] algorithm, as leaving it unfenced may cause a privacy le
       1. Set |url| to |config|'s [=fenced frame config/mapped url=]'s [=mapped url/value=].
 
       1. Run steps in |config|'s [=fenced frame config/on navigate callback=].
+  
+    1. If |navigable| is a [=fenced navigable container/fenced navigable=], and
+       <var ignore>sourceDocument</var>'s [=node navigable=] is not |navigable|:
```

Comment:
This is to limit the condition to embedder-initiated navigations, right? But this wouldn't catch navigations initiated by a deep descendant _inside_ a fenced frame, that navigates `window.top`, right? That would pass this condition, which I think is not what we want, right?

Separately, can you point me to where, in the implementation, this check is taking place? (Just so I know what I'm talking about).

Reply:
> That would pass this condition, which I think is not what we want, right?

Yes, that would be allowed under the current algorithm state, which we don't want. Should we also be concerned about the "If url is a [urn uuid](https://www.ietf.org/rfc/rfc4122.txt#urn-uuid) and navigable is a [fenced navigable](https://github.com/WICG/fenced-frame/pull/183#fenced-navigable-container-fenced-navigable):" part as well? If a descendant navigable somehow knew a valid `urn:uuid` in the fenced frame mapping, could it also navigate the fenced frame root using that `urn:uuid`? I'm not sure if there are other safeguards in place to prevent this from happening, or if we should add a "embedder-initiated" check there as well.

> Separately, can you point me to where, in the implementation, this check is taking place? (Just so I know what I'm talking about).

The impl equivalent takes place [in the navigation request file](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;l=1712-1715;drc=f522344e45882da4c7f7cb1b3a0a7bd747d654bb). It's checking for [is_embedder_initiated_fenced_frame_navigation](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;drc=f522344e45882da4c7f7cb1b3a0a7bd747d654bb;bpv=1;bpt=1;l=1663) which is set by the [`FencedFrame::Navigate()` method](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/fenced_frame/fenced_frame.cc;l=182;drc=bd15364a662c095e05f2ded9fb7c578e46281fb0). That can only be invoked from the `HTMLFencedFrameElement`.

If we want to perfectly match that, we could add an optional boolean to the [`Navigate` algorithm](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate) for `is embedder initiated fenced frame navigation`.

Alternatively, we could just check for whether `sourceDocument's node navigable` is in the `|navigable|`'s [`active document`](https://html.spec.whatwg.org/multipage/document-sequences.html#nav-document)'s [`ancestor navigables`](https://html.spec.whatwg.org/multipage/document-sequences.html#ancestor-navigables) list. I'll do this method, but if you feel like we should do the other one let me know.

Reply:
> I'll do this method, but if you feel like we should do the other one let me know.

I think that less-invasive approach makes sense. Plus I think it solves the earlier question you asked, right?

> Should we also be concerned about the "If url is a [urn uuid](https://www.ietf.org/rfc/rfc4122.txt#urn-uuid) and navigable is a https://github.com/WICG/fenced-frame/pull/183#fenced-navigable-container-fenced-navigable:" part as well? If a descendant navigable somehow knew a valid urn:uuid in the fenced frame mapping, could it also navigate the fenced frame root using that urn:uuid?

That is, your approach stops descendant navigables from doing this, right? I guess the spec could accidentally allow some descendant navigable to still navigate to a URN that it somehow guesses correctly (since it should be buried inside of a `FencedFrameConfig` object), and still cause this outcome which isn't great. Maybe as a follow-up we can make the navigate algorithm take a `FencedFrameConfig` object, to make it clear that the ability to call navigate with a URL is tied to the fenced frame navigation path only. Let me know if you think that sounds like the right approach, and then maybe you could file a follow-up issue?

Reply:
> That is, your approach stops descendant navigables from doing this, right?

Correct. I think as the spec currently stands, if a fenced frame can generate a FencedFrameConfig, figure out what the urn:uuid for it is, it can navigate **itself** to that config via `location.href = urn_i_generated`, which I don't think we want.

I'll also update the other branch to do the ancestor check as well.

While I was doing this, I realized that we don't actually have unfenced versions of the ancestor navigables algorithms. We do have them for the descendant navigables algorithms. I've updated this PR to add those patches in and call the unfenced versions of them in these 2 spots.

> Maybe as a follow-up we can make the navigate algorithm take a FencedFrameConfig object

I think that would be good. Then the thing I described above with a fenced frame navigating itself to a urn:uuid wouldn't be possible. I created https://github.com/WICG/fenced-frame/issues/194 to track this.

Reply:
> Correct. I think as the spec currently stands, if a fenced frame can generate a FencedFrameConfig, figure out what the urn:uuid for it is, it can navigate itself to that config via location.href = urn_i_generated, which I don't think we want.

This is weird. I assumed the given spec's limitation (which you've captured in #194) only made it possible for an embedder to navigate a fenced frame if it "guesses" the `urn uuid` inside the config that it already has, but it seems like the spec might also be vulnerable to a fenced frame generating a config and navigating _itself_ to that config's urn (assuming it can guess the urn of course)? I guess that's because [here in the spec](https://wicg.github.io/fenced-frame/#ref-for-traversable-navigable-fenced-frame-config-mapping%E2%91%A2) we only check the source document's traversable navigable's fenced frame config mapping.

My understanding is that the implementation does not have this limitation though, right? That's for fenced frame navigations, the [implementation just blindly consults its _embedder_'s frame tree's mapping](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;l=2525-2533;drc=e66b343b5554785a32ad988bfe5c3c524f5e1857). Do we know why the spec blindly looks at _sourceDocument_ — which might not be the embedder, but might actually be the fenced frame itself — when consulting the mapping?

I'm not sure this is true, because when we compare urn UUIDs, we always compare the given one with ones in the map of the embedder, not a frame's own map, right? (At least I think that's how the implementation works... we always look at the embedder's frame tree, and I feel like we do the same in spec?

Reply:
> Do we know why the spec blindly looks at sourceDocument — which might not be the embedder, but might actually be the fenced frame itself — when consulting the mapping?

If I were to take a guess, maybe it's because we needed access to a document at that point, and `sourceDocument` at the time felt like it was enough of an analog. At this point we can just as easily get navigable's unfenced parent's active document, right?

> I'm not sure this is true, because when we compare urn UUIDs, we always compare the given one with ones in the map of the embedder, not a frame's own map, right? (At least I think that's how the implementation works... we always look at the embedder's frame tree, and I feel like we do the same in spec?

I believe that's what's happening in the implementation. The urn lives in a mapping that lives in the embedder (at this point, the fenced frame might not even have been created yet). The spec should do that as well.

Reply:
> At this point we can just as easily get navigable's unfenced parent's active document, right?

I think, so? Assuming that doesn't break URN iframes, which I think it should be fine since the URN appears in the parent's (fenced, though)'s config mapping. Anyways, we don't have to take care of that in this PR. It can be done separately.

Reply:
Sounds good. I've linked this comment thread in https://github.com/WICG/fenced-frame/issues/194 as a reference, and that can be taken care of as part of that effort.